modules/qtwidgets/properties/listpropertywidgetqt.cpp

```diff
@@ -0,0 +1,188 @@
+
+#include <modules/qtwidgets/properties/listpropertywidgetqt.h>
+#include <modules/qtwidgets/inviwoqtutils.h>
+#include <modules/qtwidgets/inviwowidgetsqt.h>
+
+#include <inviwo/core/properties/compositeproperty.h>
+#include <inviwo/core/properties/listproperty.h>
+#include <inviwo/core/util/zip.h>
+
+#include <warn/push>
+#include <warn/ignore/all>
+#include <QMenu>
+#include <QAction>
+#include <QToolButton>
+#include <QHBoxLayout>
+#include <warn/pop>
+
+namespace inviwo {
+
+ListPropertyWidgetQt::ListPropertyWidgetQt(ListProperty* property)
+    : CompositePropertyWidgetQt(property), listProperty_(property) {
+
+    setShowIfEmpty(true);
+
+    // inject "add" button in the header of the composite property for adding list elements
+    if (auto headerlayout = dynamic_cast<QHBoxLayout*>(layout()->itemAt(0)->layout())) {
+        addItemButton_ = new QToolButton(this);
+        QIcon icon;
+        icon.addFile(":/stylesheets/images/addlistitem.png", QSize(), QIcon::Normal, QIcon::Off);
+        icon.addFile(":/stylesheets/images/addlistitem_disabled.png", QSize(), QIcon::Disabled,
+                     QIcon::Off);
+        addItemButton_->setIcon(icon);
+        addItemButton_->setObjectName("addListItemButton");
+        addItemButton_->setToolTip("Add list element");
+
+        connect(addItemButton_, &QToolButton::clicked, this, [&]() {
+            if (listProperty_ && listProperty_->getPrefabCount() > 0) {
+                if (listProperty_->getPrefabCount() == 1) {
+                    listProperty_->addProperty(static_cast<size_t>(0));
+                } else {
+                    // show context menu since there exist more than one prefab in the list property
+                    QMenu m;
+                    for (auto&& item : util::enumerate(listProperty_->getPrefabs())) {
+                        auto&& ind = get<0>(item);
+                        auto&& elem = get<1>(item);
+
+                        auto addAction = m.addAction(utilqt::toQString(elem->getDisplayName()));
+                        addAction->setData(static_cast<uint>(ind));
+                    }
+
+                    if (auto selectedAction = m.exec(QCursor::pos())) {
+                        listProperty_->setInitiatingWidget(this);
+                        listProperty_->addProperty(selectedAction->data().toUInt());
+                        listProperty_->clearInitiatingWidget();
+                    }
+                }
+            }
+        });
+
+        if (listProperty_) {
+            addItemButton_->setVisible((listProperty_->getUIFlags() & ListPropertyUIFlag::Add) ==
+                                       ListPropertyUIFlag::Add);
+
+            // listProperty_->PropertyOwnerObservable::addObserver(this);
+        }
+
+        headerlayout->insertWidget(headerlayout->count() - 2, addItemButton_);
+    }
+}
+
+void ListPropertyWidgetQt::updateFromProperty() {
+    CompositePropertyWidgetQt::updateFromProperty();
+
+    if (listProperty_) {
```

Comment:
Can listproperty be null? If so why?

Reply:
That way at least you get an issue when you change the compositepropertywidget...